vizro-core/examples/kpi/app.py

```diff
@@ -0,0 +1,271 @@
+"""Example to show dashboard configuration."""
+
+import pandas as pd
+import vizro.models as vm
+from utils._charts import COLUMN_DEFS, KPI, bar, chloropleth, line, pie
+from utils._helper import clean_data_and_add_columns
+from vizro import Vizro
+from vizro.actions import filter_interaction
+from vizro.tables import dash_ag_grid
+
+# DATA --------------------------------------------------------------------------------------------
+df_complaints = pd.read_csv("data/Financial Consumer Complaints.csv")
+df_complaints = clean_data_and_add_columns(df_complaints)
+
+vm.Page.add_type("components", KPI)
+vm.Container.add_type("components", KPI)
+
+# TODO: Enable selection of year filter
+# TODO: Enable current year vs. past year comparison
+# TODO: Enable dynamic KPI Cards
+# TODO: Make KPI Card row horizontally scrollable if no space
+# TODO: Bar - How to enable drill-downs for Issue/Sub-issue and Product/Sub-product?
+# TODO: Bar - Reformat numbers with commas in bar chart
+# TODO: Bar - Left-align y-axis labels
+# TODO: Bar - Shorten labels
+# TODO: Line - Customize function to always show selected year vs. past year
+
+
+page_exec = vm.Page(
+    title="Executive View",
+    layout=vm.Layout(
+        grid=[
+            [0, 1, 2, 3, 4, 5],
+            [0, 1, 2, 3, 4, 5],
+            [6, 6, 6, 7, 7, 7],
+            [6, 6, 6, 7, 7, 7],
+            [6, 6, 6, 7, 7, 7],
+            [6, 6, 6, 7, 7, 7],
+            [6, 6, 6, 8, 8, 8],
+            [6, 6, 6, 8, 8, 8],
+            [6, 6, 6, 8, 8, 8],
+            [6, 6, 6, 8, 8, 8],
```

Comment:
This is also a BIG variable `page_exec` already - I'd consider breaking it down a bit into more manageable chunks even with the current level of nesting. So you'd maybe have `kpi_cards` or `tabs` or similar extracted out as variables.

Reply:
P.S. I feel like I made the "factorise layout" comment a few times, and maybe that's because it's something that is only obvious to me 😅 Is this something that is not intuitive/hard to remember in general? Should we explain more in docs? Or you think it's fine and you just didn't think of it here?

Reply:
No, it's very intuitive and definitely clear to me haha It's historically here, and I just didn't refactor this part yet. 

For context, so you understand where I am coming from: When I start to develop dashboards and have lots of components/charts on the page, I usually start with a bigger number of rows as it's easier for me to control the px sizes. This page previously had a lot more charts/components, so I've started off with 10-12 rows. With the KPI cards, I've had to iterate a LOT, as I previously didn't have CSS that dynamically changes font sizes depending on the screen width/height. So I've been adding/removing rows back and forth. What you see here is the outcome of all my iterations over time, but since then I've been removing the majority of the charts/components + I've added dynamic CSS 😄 So this grid layout looks way more complicated then it needs to be, I agree 😄

So from a developing point of view, I like to start off with bigger grids as it's easier for me to control px sizes and because I don't know from the beginning what ratios look well with the components I have. I also don't know from the beginning what charts/components I want to have in, so I am replacing back and forth as well. And then after fixing the number of charts/components, I usually go back to the grid and simplify the configuration. I didn't do it here obviously as there was a bit of time pressure on Friday to get the code out and deploy 😄

Reply:
Cool, that all makes sense, thank you very much for the explanation! 🙏  At some point I think we should have development tips like this in our docs so that people can see "in reality" how you can develop a vizro dashboard. It would show different ways of working (e.g. Jupyter, IDE next to web browser with hot reloading) and tips and tricks like this.